File: talor-gui/src/api/task.ts
/**
 * Task API Module
 * 任务 API 模块
 *
 * Provides task management API calls for background task execution.
 */

import type { TalorClient } from './client';

export type TaskStatus = 'pending' | 'queued' | 'running' | 'completed' | 'failed' | 'cancelled';

export interface TaskArtifact {
  path: string;
  type: string;
  updatedAt: number;
}

export interface TaskInfo {
  id: string;
  sessionId: string;
  agentId: string;
  title: string;
  status: TaskStatus;
  progress: number;
  currentAction: string | null;
  artifacts: TaskArtifact[];
  result: string | null;
  error: string | null;
  createdAt: number;
  updatedAt: number;
  startedAt: number | null;
  completedAt: number | null;
}

export interface TaskStatusInfo {
  taskId: string;
  status: TaskStatus;
  progress: number;
  currentAction: string | null;
}

export interface CreateTaskRequest {
  title: string;
  agent_id: string;
  prompt: string;
  use_worktree?: boolean;
}

export interface WorkspaceInfo {
  task_id: string;
  worktree_path: string | null;
  artifacts: Array<{ path: string; type: string; updated_at: number }>;
  git_diff_stat: string | null;
}

export interface TaskApi {
  create(request: CreateTaskRequest): Promise<TaskInfo>;
  list(status?: TaskStatus): Promise<TaskInfo[]>;
  get(taskId: string): Promise<TaskInfo>;
  getStatus(taskId: string): Promise<TaskStatusInfo>;
  cancel(taskId: string): Promise<void>;
  getWorkspace(taskId: string): Promise<WorkspaceInfo>;
  previewFile(taskId: string, path: string): Promise<string>;
}

// Backend snake_case response
interface BackendTaskResponse {
  task_id: string;
  session_id: string;
  agent_id: string;
  title: string;
  status: TaskStatus;
  progress: number;
  current_action: string | null;
  artifacts: Array<{ path: string; type: string; updated_at: number }>;
  result: string | null;
  error: string | null;
  created_at: number;
  updated_at: number;
  started_at: number | null;
  completed_at: number | null;
}

interface BackendTaskStatusResponse {
  task_id: string;
  status: TaskStatus;
  progress: number;
  current_action: string | null;
}

function toTaskInfo(r: BackendTaskResponse): TaskInfo {
  return {
    id: r.task_id,
    sessionId: r.session_id,
    agentId: r.agent_id,
    title: r.title,
    status: r.status,
    progress: r.progress,
    currentAction: r.current_action,
    artifacts: r.artifacts.map((a) => ({
      path: a.path,
      type: a.type,
      updatedAt: a.updated_at,
    })),
    result: r.result,
    error: r.error,
    createdAt: r.created_at,
    updatedAt: r.updated_at,
    startedAt: r.started_at,
    completedAt: r.completed_at,
  };
}

export function createTaskApi(client: TalorClient): TaskApi {
  return {
    async create(request: CreateTaskRequest): Promise<TaskInfo> {
      const response = await client.post<BackendTaskResponse>('/api/tasks', request);
      return toTaskInfo(response);
    },

    async list(status?: TaskStatus): Promise<TaskInfo[]> {
      let endpoint = '/api/tasks';
      if (status) {
        endpoint += `?status=${encodeURIComponent(status)}`;
      }
      const response = await client.get<BackendTaskResponse[]>(endpoint);
      return response.map(toTaskInfo);
    },

    async get(taskId: string): Promise<TaskInfo> {
      const response = await client.get<BackendTaskResponse>(`/api/tasks/${encodeURIComponent(taskId)}`);
      return toTaskInfo(response);
    },

    async getStatus(taskId: string): Promise<TaskStatusInfo> {
      const response = await client.get<BackendTaskStatusResponse>(
        `/api/tasks/${encodeURIComponent(taskId)}/status`
      );
      return {
        taskId: response.task_id,
        status: response.status,
        progress: response.progress,
        currentAction: response.current_action,
      };
    },

    async cancel(taskId: string): Promise<void> {
      await client.post<void>(`/api/tasks/${encodeURIComponent(taskId)}/cancel`, {});
    },

    async getWorkspace(taskId: string): Promise<WorkspaceInfo> {
      return client.get<WorkspaceInfo>(`/api/tasks/${encodeURIComponent(taskId)}/workspace`);
    },

    async previewFile(taskId: string, path: string): Promise<string> {
      return client.get<string>(
        `/api/tasks/${encodeURIComponent(taskId)}/workspace/preview?path=${encodeURIComponent(path)}`
      );
    },
  };
}

export default createTaskApi;

File: talor/src/task/task.py
"""Task entity — background task data model."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from enum import Enum


class TaskStatus(str, Enum):
    PENDING = "pending"
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


@dataclass
class TaskArtifact:
    path: str           # relative path within worktree/workspace
    type: str           # "file" | "directory"
    updated_at: int     # Unix ms


@dataclass
class Task:
    id: str
    session_id: str
    agent_id: str
    title: str
    status: TaskStatus = TaskStatus.PENDING
    progress: int = 0
    current_action: str | None = None
    artifacts: list[TaskArtifact] = field(default_factory=list)
    checkpoint_path: str | None = None
    worktree_path: str | None = None
    result: str | None = None
    error: str | None = None
    created_at: int = 0
    updated_at: int = 0
    started_at: int | None = None
    completed_at: int | None = None

    def to_db_row(self) -> dict:
        return {
            "id": self.id,
            "session_id": self.session_id,
            "agent_id": self.agent_id,
            "title": self.title,
            "status": self.status.value,
            "progress": self.progress,
            "current_action": self.current_action,
            "artifacts": json.dumps([
                {"path": a.path, "type": a.type, "updated_at": a.updated_at}
                for a in self.artifacts
            ]),
            "checkpoint_path": self.checkpoint_path,
            "worktree_path": self.worktree_path,
            "result": self.result,
            "error": self.error,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
            "started_at": self.started_at,
            "completed_at": self.completed_at,
        }

    @classmethod
    def from_db_row(cls, row: dict) -> "Task":
        artifacts_raw = json.loads(row.get("artifacts") or "[]")
        artifacts = [
            TaskArtifact(
                path=a["path"],
                type=a.get("type", "file"),
                updated_at=a.get("updated_at", 0),
            )
            for a in artifacts_raw
        ]
        return cls(
            id=row["id"],
            session_id=row["session_id"],
            agent_id=row["agent_id"],
            title=row["title"],
            status=TaskStatus(row["status"]),
            progress=row.get("progress", 0) or 0,
            current_action=row.get("current_action"),
            artifacts=artifacts,
            checkpoint_path=row.get("checkpoint_path"),
            worktree_path=row.get("worktree_path"),
            result=row.get("result"),
            error=row.get("error"),
            created_at=row.get("created_at", 0) or 0,
            updated_at=row.get("updated_at", 0) or 0,
            started_at=row.get("started_at"),
            completed_at=row.get("completed_at"),
        )
